### PR TITLE
clarify namespacing

### PR DESCRIPTION
**Pull request by DavyCats**
This PR aims to clarify the namespacing of calls to workflows. Currently miniwdl and cromwell handle this differently (at least in the input JSON; see also #329), so some clarification seems to be needed in the spec. 

Pinging @mlin and @cjllanwarne as engine implementers.

### CODE REVIEW

**Comment by mlin** [comment]
Hi @DavyCats, thanks for pushing this forward! To be clear, is this saying the JSON key for a subworkflow input should *not* include the subworkflow name following the call name (as Cromwell currently expects)? If so, I like that on first principles but we probably need to talk about how breaking this is for Cromwell and existing input JSON files.

**Comment by patmagee** [comment]
@DavyCats I like this as well. The namespacing has always felt a little weird. There are a considerable number of breaking changes currently in the propsosed spec, I wonder if we could do this in a non-breaking way: ie Propose that `<wf>.<task>.<subworkflow>.<input>` is actually an alias for `<wf>.<task>.<input>` where the optional `<subworkflow>` will be deprecated in a future version of the WDL specification. @mlin  @cjllanwarne as implementors would this feel better?

**Comment by mlin** [comment]
Generally sounds reasonable to me. There might be some dumb corner cases to worry about, for example if the subworkflow has a call given the same name as the subworkflow itself:

* lib.wdl:
```
version 1.0

workflow subsubwf {
  input {
    Int bar
  }
}
```
* subwf.wdl:
```
version 1.0

import "lib.wdl"

workflow subwf {
  input {
    Int bar
  }

  call lib.subsubwf as subwf
}
```

Then with the aliasing, `wf.callname.subwf.bar` would be an ambiguous reference. I'm not saying this dumb case should shoot down the overall proposal, just that we'd need a rule for resolving it.

**Comment by DavyCats** [comment]
@mlin I feel like that should be a name collision:
> "[...]The names of the contained namespaces, tasks, and workflow need to be unique within that namespace. For example, one cannot import two workflows while they have the same namespace identifier. Additionally, a workflow and a namespace both named `foo` cannot exist inside a common namespace. Similarly there cannot be a task `foo` in a workflow also named `foo`. [...]"
>
> last part of https://github.com/openwdl/wdl/blob/master/versions/development/SPEC.md#namespaces

**Comment by patmagee** [comment]
Hey @mlin, in the example you gave, wouldnt `wf.subwf.subsubwf.bar` always evauluate to the `bar` input of `subsubwf` even when you dealias and turn it back into`wf.subwf.bar`?

**Comment by DavyCats** [comment]
@patmagee I made some adjustments based on your comment. Is this what you had in mind?

**Comment by patmagee** [comment]
@DavyCats that makes alot sense to me. @mlin can you verify that this works?

**Comment by mlin** [comment]
My crappy example https://github.com/openwdl/wdl/pull/340#issuecomment-553092014 currently validates with both womtool and miniwdl. I'm open to simply declaring them both in spec violation :sunglasses: The section @DavyCats cited might be clarified

> Similarly there cannot be a ~~task~~ **call** foo in a workflow also named foo

Def interested in Cromwell team input here

**Comment by rhpvorderman** [comment]
Good work @DavyCats .

> The names of the contained namespaces, tasks, and workflow need to be unique within that namespace.

Should we also specify inputs here? For example

```wdl
version 1.0

import foo_workflows.wdl as foo_wf
workflow DirtyNamespace {
    input {
        String foo
    }
    call foo_wf.foo as foo {
          input:
               foo=foo
     }
}
```
`DirtyNamespace.foo` as both a call and an input here. So how does that work according to the spec?

**Comment by DavyCats** [comment]
After some consideration of @rhpvorderman's comment and reading the namespaces section a few times, I found that it was a bit hard to follow (it doesn't clearly define what a namespace is) and thus decided to rewrite it entirely. It also contained some notes about imports, which seemed redundant and out of place, which I  have now removed.

My intend is for the specifications to be the same as before, just spelled out a bit more clearly. Please let me know if this should be changed (or reverted).

**Comment by mlin** [comment]
https://github.com/chanzuckerberg/miniwdl/pull/309

There's a branch on miniwdl which (i) prevents calls from having the same name as their containing workflow and (ii) attempts to delete the subworkflow name from the input JSON keys. Seems OK so far

**Comment by cjllanwarne** [comment]
Sorry that I didn't spot this PR earlier - what you found in Cromwell looks like a bug rather than intended design since Cromwell (and WDL!) never intentionally allowed passing inputs through to subworkflow calls from input json.

Since #347 is addressing that - and in theory we'll be "adding" it for the first time, I'd be inclined to remove the "backwards-compatible-with-a-bug" option and define this as the only logical (at least to me...) syntax of `wf.call.subcall.subcall.inputname`.

**Comment by DavyCats** [comment]
Thanks @cjllanwarne, I guess it was a common misconception that inputs could be passed to sub workflows through the inputs json. I have removed the "backwards-compatible-with-a-bug" option.

**Comment by rhpvorderman** [comment]
Can #347 be closed if this is accepted? I think so. @DavyCats what do you think?

Also. I think we should get this into voting. I believe Cromwell 48 is waiting on a decision to support namespaced inputs in https://github.com/broadinstitute/cromwell/pull/5317. I believe @cjllanwarne coded the input JSON parsing to easily allow support for namespaced inputs on purpose. But currently it is not turned on because the current development spec does not allow it.

**Comment by DavyCats** [comment]
I'm not sure, the changes in #347 should still be made I think but we could fold it into this PR. This PR doesn't (in its current state) alter the fact that the spec explicitly disallows the provision of inputs to subworkflows by the end-user.

**Comment by rhpvorderman** [comment]
Well in that case we can also take the time to carefully evaluate this PR and make sure #347 is merged quickly.

**Comment by patmagee** [comment]
@rhpvorderman is this ready for voting?

**Comment by DavyCats** [comment]
@patmagee Yes, I believe so

**Comment by aednichols** [comment]
👍

**Comment by patmagee** [comment]
@mlin @cjllanwarne @jdidion  before I pass this and merge, can you please give your final thoughts on it

File: versions/development/SPEC.md
Comment (jdidion) [inline versions/development/SPEC.md:188]: s/should/must/

File: versions/development/SPEC.md
Comment (jdidion) [inline versions/development/SPEC.md:2287]: add link to relevant place in the spec

File: versions/development/SPEC.md
Comment (jdidion) [inline versions/development/SPEC.md:2290]: add link to relevant place in the spec

**Review by jdidion — APPROVED** [review]

**Comment by jdidion** [comment]
👍 with minor comments

**Comment by mlin** [comment]
LGTM, my only comment re @jdidion on the JSON is that the first namespace -- the name of the top-level workflow being run -- doesn't constrain or disambiguate anything, so I think it's nice to leave a little wiggle room to omit it from the JSON keys.

**Comment by DavyCats** [comment]
I addressed your comments @jdidion.

**Comment by jdidion** [comment]
This is in v1.1 and will be propagated to development.

**Pull request closed without merging**
